src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs

```diff
@@ -0,0 +1,1864 @@
+        }
+
+        /// <summary>Represents a key/value pair in the dictionary.</summary>
+        private struct Entry
+        {
+            /// <summary>The index of the next entry in the chain, or -1 if this is the last entry in the chain.</summary>
+            public int Next;
+            /// <summary>Cached hash code of <see cref="Key"/>.</summary>
+            public uint HashCode;
+            /// <summary>The key.</summary>
+            public TKey Key;
+            /// <summary>The value associated with <see cref="Key"/>.</summary>
+            public TValue Value;
+        }
+
+        /// <summary>Enumerates the elements of a <see cref="OrderedDictionary{TKey, TValue}"/>.</summary>
+        [StructLayout(LayoutKind.Auto)]
+        public struct Enumerator : IEnumerator<KeyValuePair<TKey, TValue>>, IDictionaryEnumerator
+        {
+            /// <summary>The dictionary being enumerated.</summary>
+            private readonly OrderedDictionary<TKey, TValue> _dictionary;
+            /// <summary>A snapshot of the dictionary's version when enumeration began.</summary>
+            private readonly int _version;
+            /// <summary>Whether Current should be a DictionaryEntry.</summary>
+            private readonly bool _useDictionaryEntry;
+            /// <summary>The current index.</summary>
+            private int _index;
+
+            /// <summary>Initialize the enumerator.</summary>
+            internal Enumerator(OrderedDictionary<TKey, TValue> dictionary, bool useDictionaryEntry)
+            {
+                _dictionary = dictionary;
+                _version = _dictionary._version;
+                _useDictionaryEntry = useDictionaryEntry;
+            }
+
+            /// <inheritdoc/>
+            public KeyValuePair<TKey, TValue> Current { get; private set; }
+
+            /// <inheritdoc/>
+            readonly object IEnumerator.Current => _useDictionaryEntry ?
+                new DictionaryEntry(Current.Key, Current.Value) :
+                Current;
+
+            /// <inheritdoc/>
+            readonly DictionaryEntry IDictionaryEnumerator.Entry => new(Current.Key, Current.Value);
+
+            /// <inheritdoc/>
+            readonly object IDictionaryEnumerator.Key => Current.Key;
+
+            /// <inheritdoc/>
+            readonly object? IDictionaryEnumerator.Value => Current.Value;
+
+            /// <inheritdoc/>
+            public bool MoveNext()
+            {
+                OrderedDictionary<TKey, TValue> dictionary = _dictionary;
+
+                if (_version != dictionary._version)
+                {
+                    ThrowHelper.ThrowVersionCheckFailed();
+                }
+
+                if (_index < dictionary._count)
+                {
+                    Debug.Assert(dictionary._entries is not null);
+                    ref Entry entry = ref dictionary._entries[_index];
+                    Current = new KeyValuePair<TKey, TValue>(entry.Key, entry.Value);
```

Comment:
Nit: looks like the ctor could be target typed?

Reply:
We've generally avoided doing that when the type isn't visible on the lhs.